### PR TITLE
Deduce sms provider config from env

### PR DESCRIPTION
**Pull request by chankiyu22**

### CODE REVIEW

File: pkg/lib/config/sms_gateway.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package config
```
Comment (louischan-oursky) [inline pkg/lib/config/sms_gateway.go:1]: The convention is to name this file `sms_gateway_env.go`. For example, `redis_env.go`.

File: pkg/lib/config/sms_gateway.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+type SMSGatewayDefaultConfig struct {
+	UseConfigFrom UseConfigFrom `envconfig:"USE_CONFIG_FROM"`
+	Provider      SMSProvider   `envconfig:"PROVIDER"`
+}
+
+type SMSGatewayConfig struct {
+	Twilio  *SMSGatewayTwilioCredentials       `envconfig:"TWILIO"`
+	Nexmo   *SMSGatewayNexmoCredentials        `envconfig:"NEXMO"`
+	Custom  *SMSGatewayCustomSMSProviderConfig `envconfig:"CUSTOM"`
+	Default *SMSGatewayDefaultConfig           `envconfig:"DEFAULT"`
```
Comment (louischan-oursky) [inline pkg/lib/config/sms_gateway.go:35]: The convention of env config struct is do not use pointers.

File: pkg/lib/config/sms_gateway.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+type SMSGatewayNexmoCredentials struct {
+	APIKey    string `envconfig:"API_KEY"`
+	APISecret string `envconfig:"API_SECRET"`
+}
+
+type SMSGatewayCustomSMSProviderConfig struct {
+	URL     string           `envconfig:"URL"`
+	Timeout *DurationSeconds `envconfig:"TIMEOUT"`
```
Comment (louischan-oursky) [inline pkg/lib/config/sms_gateway.go:16]: The convention of env config struct is do not use pointers.

File: pkg/lib/config/sms_gateway.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package config
+
+type SMSGatewayTwilioCredentials struct {
```
Comment (louischan-oursky) [inline pkg/lib/config/sms_gateway.go:3]: Change the prefix of alls struct in this file to `SMSGatewayEnvironment`, for example, `SMSGatewayEnvironmentTwilioCredentials`

File: pkg/lib/config/messaging.go [hunk 5]
```diff
@@ -16,6 +17,7 @@ var _ = Schema.Add("MessagingConfig", `
 
 type MessagingConfig struct {
 	SMSProvider      SMSProvider                `json:"sms_provider,omitempty"`
+	SMSGateway       *SMSGateway                `json:"sms_gateway,omitempty" nullable:"true"`
```
Comment (louischan-oursky) [inline pkg/lib/config/messaging.go:20]: Rename the struct to `SMSGatewayConfig`, to stick with the naming convention.

File: pkg/lib/infra/sms/config_resolver.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+import (
+	"fmt"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+)
+
+type ConfigResolver struct {
+	AuthgearSecretsYAMLSMSProvider config.SMSProvider
+	AuthgearSecretsYAMLSMSGateway  *config.SMSGateway
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/config_resolver.go:11]: The name is confusing. Do you mean `AuthgearYAMLSMSProvider`, `AuthgearYAMLSMSGateway`?

File: pkg/lib/infra/sms/config_resolver.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	return r.resolveProviderConfig(
+		r.AuthgearSecretsYAMLCustomSMSProviderConfig,
+		(*config.CustomSMSProviderConfig)(r.EnvironmentCustomSMSProviderConfig),
+	).(*config.CustomSMSProviderConfig)
+}
+
+func (r *ConfigResolver) resolveProviderConfig(
+	configFromAuthgearSecretsYAML interface{},
+	configFromEnv interface{},
+) interface{} {
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/config_resolver.go:92]: The signature is not type-strong. Can you use generics here?

File: pkg/lib/infra/sms/provider_resolver_test.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+		}
+
+		type EnvConfig struct {
+			SMSGateway *config.SMSGatewayConfig `envconfig:"SMS_GATEWAY"`
+		}
+
+		type TestCase struct {
+			Name                 string      `yaml:"name"`
+			Result               string      `yaml:"result"`
+			AuthgearSecretsYAML  interface{} `yaml:"authgear.secrets.yaml"`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/provider_resolver_test.go:42]: The testcase should take `authgear.yaml`, `authgear.secrets.yaml`, and `environment_variables`.

File: pkg/lib/infra/sms/config_resolver.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
+package sms
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/config_resolver.go:1]: I suggest we have client_resolver.go only. You unify config_resolver and provider_resolver and make them client_resolver.

File: pkg/lib/infra/sms/client.go [hunk 10]
```diff
@@ -38,6 +38,7 @@ type Client struct {
 	TwilioClient                 *TwilioClient
 	NexmoClient                  *NexmoClient
 	CustomClient                 *CustomClient
+	ProviderResolver             *ProviderResolver
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client.go:40]: 1. Remove TwilioClient, NexmoClient, CustomClient in this struct
2. Remove NewTwilioClient, NewNexmoClient, NewCustomClient from the deps.
3. Inject ClientResolver here.
4. Use ClientResolver to resolve a client to use.
5. The API of ClientResolver is `func ResolveClient() (RawClient, error)`

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by chankiyu22** [comment]
@louischan-oursky Updated.

File: pkg/lib/config/sms_gateway_env.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	APIKey    string `envconfig:"API_KEY"`
+	APISecret string `envconfig:"API_SECRET"`
+}
+
+type SMSGatewayEnvironmentCustomSMSProviderConfig struct {
+	URL     string          `envconfig:"URL"`
+	Timeout DurationSeconds `envconfig:"TIMEOUT"`
+}
+
+type SMSGatewayEnvironmentUseConfigFrom string
```
Comment (louischan-oursky) [inline pkg/lib/config/sms_gateway_env.go:19]: Add type SMSGatewayEnvironmentDefaultProvider string

Comment (louischan-oursky) [inline pkg/lib/config/sms_gateway_env.go:19]: Rename SMSGatewayEnvironmentUseConfigFrom to SMSGatewayEnvironmentDefaultUseConfigFrom

File: pkg/lib/infra/sms/client_resolver.go [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+
+type ClientResolver struct {
+	AuthgearYAMLSMSProvider config.SMSProvider
+	AuthgearYAMLSMSGateway  *config.SMSGatewayConfig
+
+	AuthgearSecretsYAMLNexmoCredentials        *config.NexmoCredentials
+	AuthgearSecretsYAMLTwilioCredentials       *config.TwilioCredentials
+	AuthgearSecretsYAMLCustomSMSProviderConfig *config.CustomSMSProviderConfig
+
+	EnvironmentDefaultProvider      config.SMSProvider
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:17]: Use the new type `config.SMSGatewayEnvironmentDefaultProvider`

File: pkg/lib/infra/sms/client_resolver.go [hunk 13]
```diff
@@ -25,6 +25,31 @@ type ClientResolver struct {
 	SMSWebHook  SMSWebHook
 }
 
+func ProvideClientResolver(
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:28]: Remove this

File: pkg/lib/infra/sms/client_resolver.go [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+	return client, smsClientCredentials, nil
+}
+
+func (r *ClientResolver) resolveProvider() config.SMSProvider {
+	if r.AuthgearYAMLSMSGateway != nil {
+		// Use sms gateway config. See Table 3
+		return r.resolveProviderFromAuthgearSecretsYAML()
+	}
+	if r.AuthgearYAMLSMSProvider != "" {
+		// Use `messaging.sms_provider` from `authgear.secrets.yaml`. Read config from `sms.{messaging.sms_provider}` from `authgear.secrets.yaml`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:132]: `authgear.yaml`

File: pkg/lib/infra/sms/client_resolver.go [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+	}
+	// sms_provider == "" and sms_gateway == nil
+	// See table 2
+	return r.resolveProviderFromEnv()
+}
+
+// Table 2
+func (r *ClientResolver) resolveProviderFromEnv() config.SMSProvider {
+	if r.EnvironmentDefaultUseConfigFrom == "" {
+		// `provider` will be determined from application logic. Read config from `sms.{provider}` from `authgear.secrets.yaml`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:143]: `authgear.yaml`

File: pkg/lib/infra/sms/client_resolver.go [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+	switch r.EnvironmentDefaultUseConfigFrom {
+	case config.SMSGatewayEnvironmentUseConfigFromEnvironmentVariable:
+		if r.EnvironmentDefaultProvider == "" {
+			// `provider` will be determined from application logic. Read config from `SMS_GATEWAY_{provider}_*` from environment variables
+			return ""
+		}
+		// Use `SMS_GATEWAY_DEFAULT_PROVIDER` as provider. Will read config from `SMS_GATEWAY_{SMS_GATEWAY_DEFAULT_PROVIDER}_*` environment variables
+		return r.EnvironmentDefaultProvider
+	case config.SMSGatewayEnvironmentUseConfigFromAuthgearSecretsYAML:
+		// `provider` will be determined from application logic. Read config from `sms.{provider}` from `authgear.secrets.yaml`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:155]: `authgear.yaml`

File: pkg/lib/infra/sms/client_resolver.go [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+				return ""
+			}
+			// Use `SMS_GATEWAY_DEFAULT_PROVIDER` as provider. Will read config from `SMS_GATEWAY_{SMS_GATEWAY_DEFAULT_PROVIDER}_*` environment variables
+			return r.EnvironmentDefaultProvider
+		}
+		// Use `sms_gateway.provider` as provider. Will read config from `SMS_GATEWAY_{sms_gateway.provider}_*` environment variables
+		return r.AuthgearYAMLSMSGateway.Provider
+	case config.SMSGatewayUseConfigFromAuthgearSecretsYAML:
+		// `sms_gateway.provider` is required
+		// Use provider configs from `authgear.secrets.yaml`. Will read config from `sms.{sms_gateway.provider}` from `authgear.secrets.yaml`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:179]: `authgear.yaml`

File: pkg/lib/infra/sms/client_resolver.go [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+	case config.SMSGatewayEnvironmentUseConfigFromAuthgearSecretsYAML:
+		// `provider` will be determined from application logic. Read config from `sms.{provider}` from `authgear.secrets.yaml`
+		return r.clientsFromAuthgearSecretsYAML()
+	default:
+		panic(fmt.Errorf("Invalid DEFAULT_USE_CONFIG_FROM %v", r.EnvironmentDefaultUseConfigFrom))
+	}
+}
+
+// Table 3
+func (r *ClientResolver) resolveConfigFromAuthgearSecretsYAML() (*NexmoClient, *NexmoClientCredentials, *TwilioClient, *TwilioClientCredentials, *CustomClient, *CustomClientCredentials) {
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:283]: FromAuthgearYAMLAndAuthgearSecretsYAML

File: pkg/lib/infra/sms/client_resolver.go [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+				return r.clientsFromEnv()
+			}
+			// Use `SMS_GATEWAY_DEFAULT_PROVIDER` as provider. Will read config from `SMS_GATEWAY_{SMS_GATEWAY_DEFAULT_PROVIDER}_*` environment variables
+			return r.clientsFromEnv()
+		}
+		// Use `sms_gateway.provider` as provider. Will read config from `SMS_GATEWAY_{sms_gateway.provider}_*` environment variables
+		return r.clientsFromEnv()
+	case config.SMSGatewayUseConfigFromAuthgearSecretsYAML:
+		// `sms_gateway.provider` is required
+		// Use provider configs from `authgear.secrets.yaml`. Will read config from `sms.{sms_gateway.provider}` from `authgear.secrets.yaml`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client_resolver.go:298]: `authgear.yaml`

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by chankiyu22** [comment]
@louischan-oursky updated

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**